geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java

```diff
@@ -121,78 +104,48 @@ public static void execute(ExecutablePool pool, String region, Function function
           throw se;
         }
 
-        reexecuteForServ = true;
+        isReexecute = true;
         resultCollector.clearResults();
         failedNodes.clear();
       }
-    } while (reexecuteForServ);
-
-    if (reexecute && function.isHA()) {
-      ExecuteRegionFunctionOp.reexecute(pool, region, function, serverRegionExecutor,
-          resultCollector, hasResult, failedNodes, maxRetryAttempts);
-    }
+    } while (true);
   }
 
-  public static void execute(ExecutablePool pool, String region, String function,
+  /**
+   * Does a execute Function on a server using connections from the given pool to communicate with
+   * the server.
+   *
+   * @param pool the pool to use to communicate with the server.
+   * @param region the name of the region to do the put on
+   * @param function to be executed
+   * @param serverRegionExecutor which will return argument and filter
+   * @param resultCollector is used to collect the results from the Server
+   * @param hasResult is used to collect the results from the Server
+   * @param maxRetryAttempts Maximum number of retry attempts
+   */
+  public static void execute(ExecutablePool pool, String region, Function function,
       ServerRegionFunctionExecutor serverRegionExecutor, ResultCollector resultCollector,
-      byte hasResult, int mRetryAttempts, boolean isHA, boolean optimizeForWrite) {
+      byte hasResult, int maxRetryAttempts) {
 
     AbstractOp op = new ExecuteRegionFunctionOpImpl(region, function, serverRegionExecutor,
-        resultCollector, hasResult, new HashSet<String>(), isHA, optimizeForWrite, true);
-    boolean reexecute = false;
-    boolean reexecuteForServ = false;
-    Set<String> failedNodes = new HashSet<String>();
-    AbstractOp reexecOp = null;
+        resultCollector, hasResult, new HashSet<String>());
```

Comment:
We should avoid allocating containers if we aren't going to use them. In a stable run this method will never use the `failedNodes` parameter. Perhaps passing `null` and then checking for `null` and allocating if and only if we need it.

Reply:
I assume this refactor to pass in the `failedNodes` was to check it in some tests but I don't see those tests in this PR, so what is the purpose?

Reply:
What about Collections.emptyMap()? Logically and from jit compilation perspective it is better to have single path of execution.

Reply:
The problem with using null instead of allocating a container with no elements is that the logic would have to be complicated in several places. Whenever we need to clear, add or remove elements to the collection we would have to first check if it is either null or empty and if so, allocate it first.
The same would happen with Collections.emptyMap() given that it returns an immutable object.
In the original implementation, the execute methods already allocated an empty HashSet at the begining of the execute methods.
I'd rather leave this part as it is in order not to embroil the logic.

Regarding the passing of failedNodes to the method, the idea was to have one single private method containing the logic of the current execute and reexecute public methods that was basically repeated. The logic is now in one method but in the case of the execute public method, the value of the failedNodes argument to pass to the new method would be an empty set.

Reply:
Understood, and in non-critical paths I wouldn't even raise this issue. My concern is that in 99.9% of the times we execute this operation the map will be created fo no reason. When we get to optimizing function execution this map will likely be source of transient garbage production that puts pressure on the GC as a result of premature allocation. We will likely refactor it for lazy initialization then. It would be great if you could work under that assumption now when already refactoring this code.

Reply:
I have sent a new commit with lazy allocation of failedNodes.
Hard to decide if the gain with less GC pressure is worth with respect to the added complexity in the code.

Reply:
How about @kohlmu-pivotal's suggestion, @albertogpz (to use `Collections.emptyMap()`). That would address @pivotal-jbarrett's concern and keep the code simple right?

Reply:
Woops looks like you already took @kohlmu-pivotal's advice @albertogpz 👍